hack/verify-container-images.sh

```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
+
+# Copyright 2022 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+TRIVY_VERSION=0.34.0
+
+GO_OS="$(go env GOOS)"
+if [[ "${GO_OS}" == "linux" ]]; then
+  TRIVY_OS="Linux"
+elif [[ "${GO_OS}" == "darwin"* ]]; then
+  TRIVY_OS="macOS"
+fi
+
+GO_ARCH="$(go env GOARCH)"
+if [[ "${GO_ARCH}" == "amd" ]]; then
+  TRIVY_ARCH="32bit"
+elif [[ "${GO_ARCH}" == "amd64"* ]]; then
+  TRIVY_ARCH="64bit"
+elif [[ "${GO_ARCH}" == "arm" ]]; then
+  TRIVY_ARCH="ARM"
+elif [[ "${GO_ARCH}" == "arm64" ]]; then
+  TRIVY_ARCH="ARM64"
+fi
+
+TOOL_BIN=hack/tools/bin
+
+# Downloads trivy scanner
+curl -L -o ${TOOL_BIN}/trivy.tar.gz \
+    https://github.com/aquasecurity/trivy/releases/download/v${TRIVY_VERSION}/trivy_${TRIVY_VERSION}_${TRIVY_OS}-${TRIVY_ARCH}.tar.gz \
+
+tar xfO ${TOOL_BIN}/trivy.tar.gz trivy > ${TOOL_BIN}/trivy
+chmod +x ${TOOL_BIN}/trivy
+rm ${TOOL_BIN}/trivy.tar.gz
+
+# Builds all the container images to be scanned and cleans up changes to ./*manager_image_patch.yaml ./*manager_pull_policy.yaml.
+make REGISTRY=gcr.io/k8s-staging-cluster-api PULL_POLICY=IfNotPresent TAG=dev docker-build
+make clean-release-git
+
+# Scan the images
```

Comment:
Thx, nice work!

Looked a bit around and seems like there is some prior art in the kubernetes ecosystem. Probably more but I found something in secrets-store-csi-driver: https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/Makefile#L155-L161

Seems like they show all vulnerabilities but only fail if they find an "unfixed" one. I assume unfixed means that there is no fixed version of the dependency with the vulnerability.

I think if we include test-extension we should also include CAPD.

I would suggest:
* run the image scan for all images first
* then run the image scan with --exit-code=1 only for our prod images (clusterctl, cluster-api, KCP, CABPK)

I think it would be good to add a weekly GitHub action (like the linkchecker) which runs this script (on all supported release branches).

This way we can find the vulnerabilities relatively early and easily.

I'm fine with adding this as an additional step when cutting a release. Although it will increase the manual effort during a release significantly as all images have to be build and scanned (for every release / relesae branch) which will take some time.

I think for now I would start with trying to fix the vulnerabilities when we find them. I"m not sure what we should do when we find vulnerabilities which:
1. don't have a fix yet
2. require bumps in upstream dependencies (e.g. controller-runtime, k/k)

Would we delay the release indefinitely / who can make the assessment how critical the vulnerability is?

But as I said, I'm fine with starting on a best-effort basis and then we can see how it goes. WIth the GitHub dependency scan we didn't run into big problems as we simply had to bump dependencies.

Reply:
kk, let's keep the script as it is, I will look into the weekly GitHub action as a next step

Reply:
Added the GitHub action as well